libraries/botbuilder-ai/src/luisRecognizer.ts

```diff
@@ -43,40 +45,188 @@ export class LuisRecognizer extends IntentRecognizer {
+        return recognizer.recognizeAndMap(utterance, options.verbose || true);
+    }
+
+    protected recognizeAndMap(utterance: string, verbose: boolean): Promise<RecognizerResult> {
+        let $this = this;
+        return this.luisClient.getIntentsAndEntitiesV2($this.options.appId, this.options.subscriptionKey, utterance, $this.options.options)
+            .then((result : LuisResult) => {
+                let recognizerResult  : RecognizerResult = {
+                    text: result.query,
+                    intents: $this.getIntents(result),
+                    entities: $this.getEntitiesAndMetadata(result.entities, result.compositeEntities, verbose)
+                };                
+                return recognizerResult;
+            });
     }
 
-    protected static recognizeAndMap(client: LuisClient, utterance: string, options: LuisRecognizerOptions): Promise<Intent> {
-        return client.getIntentsAndEntitiesV2(options.appId, options.subscriptionKey, utterance, options.options)
-            .then(result => {
-                var topScoringIntent = result.topScoringIntent || {intent: '', score: 0.0};
-                return {
-                    name: topScoringIntent.intent,
-                    score: topScoringIntent.score,
-                    entities: result.entities.map(entity => {
-                        return {...{value: entity.entity}, ...entity} as  EntityObject<string>;
-                    })
-                } as Intent
+    private getIntents(luisResult: LuisResult) : any {
+        let intents : any = {}
+        if(luisResult.intents){
+            luisResult.intents.reduce((prev : any, curr : Intent) => {
+                prev[curr.intent || ''] = curr.score;
+                return prev;
+            }, intents);
+        }
+        else{
+            let topScoringIntent = luisResult.topScoringIntent || {intent: '', score: 0};
+            intents[(topScoringIntent).intent || ''] = topScoringIntent.score;
+        }
+        return intents;
+    }
+
+    private getEntitiesAndMetadata(entities: Entity[], compositeEntities : CompositeEntity[] | undefined, verbose: boolean) : any {
+        let $this = this;
+
+        let entitiesAndMetadata : any = verbose ? {$instance: {} } : {};
+        let compositeEntityTypes : string[] = [];
+
+        // We start by populating composite entities so that entities covered by them are removed from the entities list
+        if(compositeEntities){
+            compositeEntityTypes = compositeEntities.map(compositeEntity => compositeEntity.parentType);
+            compositeEntities.forEach(compositeEntity => {
+                entities = $this.populateCompositeEntity(compositeEntity, entities, entitiesAndMetadata, verbose);
             });
+        }
+
+        entities.forEach(entity => {
+            // we'll address composite entities separately
+            if(compositeEntityTypes.indexOf(entity.type) > -1)
+                return;
```

Comment:
Probably a philosophical issue and does not need to change, but I prefer:
1) Always using {} rather than relying on indenting for if clauses.  
2) Don't do return in the middle of functions--it is in effect a goto where it looks like you can add code later, but don't realize it won't be executed.
These both come from writing solid code

Reply:
1. Added the {}
2. I agree in general but in this instance the return just acts as a "continue" in order to reduce indentation in the function, I think reduced indentation in short blocks makes code more readable at a low expensive of an easy to follow goto.